src/GraphQL/Execution/DocumentExecuter.cs

```diff
@@ -496,24 +499,34 @@ public object CoerceValueAst(Schema schema, GraphType type, object input, Variab
 
             if (type is ObjectGraphType || type is InputObjectGraphType)
             {
-                var objType = type;
                 var obj = new Dictionary<string, object>();
-                var dict = input as Dictionary<string, object>;
 
-                if (dict == null)
+                if (input is KeyValuePair<string, object>)
                 {
-                    return null;
+                    var kvp = (KeyValuePair<string, object>)input;
+                    input = new Dictionary<string, object>
+                    {
+                        { kvp.Key, kvp.Value },
+                    };
                 }
 
-                objType.Fields.Apply(field =>
+                var kvps = input as IEnumerable<KeyValuePair<string, object>>;
+                if (kvps != null)
                 {
-                    if (dict.ContainsKey(field.Name))
-                    {
-                        var fieldValue = CoerceValueAst(schema, schema.FindType(field.Type), dict[field.Name], variables);
-                        obj[field.Name] = fieldValue ?? field.DefaultValue;
-                    }
+                    input = kvps.ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
+                }
+
+                var dict = (Dictionary<string, object>)input;
+
+                type.Fields.Apply(field =>
+                {
+                    object inputValue;
+                    dict.TryGetValue(field.Name, out inputValue);
+                    var fieldValue = CoerceValueAst(schema, schema.FindType(field.Type), inputValue, variables);
+                    obj[field.Name] = fieldValue ?? field.DefaultValue;
                 });
-                return obj;
+
+                input = obj;
```

Comment:
Can't remember exactly - there was some reason, but it might have been an artefact from something else. Will dig into it.

I'm happy to sync these two up. I can do that as part of this PR if you want. Any reason why you kept them separate in the first place?